### PR TITLE
Implement blocked contacts setting inside profile messages settings #…

### LINKED ISSUE #19331: Improve `quo/overlay` to avoid needing `:key` props for children
# Problem

At the moment, the `quo/overlay` component does not use `into` to compose it's children components, which leads to using the `:key` prop on children components.

Here's a prior discussion about the issue: https://github.com/status-im/status-mobile/pull/18674#discussion_r1473539054

## Implementation

Ideally, the `quo/overlay` component would be adapted to use `into` when composing its children components.

## Acceptance Criteria

Ideally, the affected areas using `quo/overlay` will be reviewed and usages of the `:key` prop would be removed (if they're no longer necessary).

### LINKED ISSUE #19590: Implement blocked contacts setting inside profile messages settings
### Summary

At the moment, we've not yet implemented the Messages settings inside a user's profile settings. Ideally, we would begin implementing these settings, and we'll start by implementing the "Blocked Contacts" settings inside the profile messages settings.

[Figma designs](https://www.figma.com/file/JlpPhJp0SMnEyBQy4nOZHo/Settings-for-Mobile?type=design&node-id=6397-13208&mode=design&t=pO44DyJHJNmtAA6i-4)

### Example of Not Implemented Messages Settings

https://github.com/status-im/status-mobile/assets/2845768/f13d9eff-45c8-44a6-861b-025ad3676240

### PR DESCRIPTION
**Pull request by flexsurfer**
fixes #19590
fixes https://github.com/status-im/status-mobile/issues/19331

refactored blocking events
refactored password settings screen with latest quo components
added top-inset? to overlay components
added custom accessories to button component

### CODE REVIEW

File: src/quo/components/list_items/user.cljs [hunk 1]
```diff
@@ -17,26 +17,28 @@
...
-       :accessibility-label :user-list-toggle-check
-       :disabled?           disabled?
-       :on-change           (when on-check on-check)}]
-     :close
-     [text/text "not implemented"]
-     [rn/view])])
+(defn accessory-item
+  [{:keys [type on-press on-check disabled? checked? child]} customization-color theme]
+  (if (= type :custom)
+    child
```
Comment (flexsurfer) [inline src/quo/components/list_items/user.cljs:23]: please comment if you have better ideas :)

File: src/status_im/subs/contact.cljs [hunk 2]
```diff
@@ -1,7 +1,7 @@
 (ns status-im.subs.contact
```
Comment (flexsurfer) [inline src/status_im/subs/contact.cljs:1]: just moved code here, i didn't write it so don't review this namespace

File: src/quo/components/list_items/user.cljs (hunk 1 above)
Comment (clauxx) [reply src/quo/components/list_items/user.cljs:23]: An alternative could be having a :button type and pass the label/on-press, but we should also address this with the design team since they went beyond the design system here it seems.

File: src/quo/components/list_items/user.cljs [hunk 3]
```diff
@@ -72,4 +74,4 @@
          :style {:color (colors/theme-colors colors/neutral-50 colors/neutral-40)}}
```
Comment (clauxx) [inline src/quo/components/list_items/user.cljs]: should pass the theme here i think

File: src/status_im/contexts/profile/settings/screens/messages/blocked_users/view.cljs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            [react-native.core :as rn]
+            [status-im.common.contact-list-item.view :as contact-list-item]
+            [status-im.contexts.profile.contact.unblock-contact.view :as unblock-contact]
+            [utils.i18n :as i18n]
+            [utils.re-frame :as rf]))
+
+(defn list-item
+  [user _ _ _]
+  [contact-list-item/contact-list-item
+   {:on-press  #()
```
Comment (clauxx) [inline src/status_im/contexts/profile/settings/screens/messages/blocked_users/view.cljs]: why is this needed?

File: src/status_im/contexts/profile/settings/screens/messages/blocked_users/view.cljs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+(defn view
+  []
+  (let [contacts (rf/sub [:contacts/blocked])]
+    [quo/overlay {:type :shell :top-inset? true}
+     [quo/page-nav
+      {:key        :header
+       :background :blur
+       :icon-name  :i/arrow-left
+       :on-press   #(rf/dispatch [:navigate-back])}]
```
Comment (clauxx) [inline src/status_im/contexts/profile/settings/screens/messages/blocked_users/view.cljs]: can be defined outside the component

File: src/status_im/contexts/profile/settings/screens/messages/view.cljs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+            [utils.i18n :as i18n]
+            [utils.re-frame :as rf]))
+
+(defn view
+  []
+  [quo/overlay {:type :shell :top-inset? true}
+   [quo/page-nav
+    {:background :blur
+     :icon-name  :i/arrow-left
+     :on-press   #(rf/dispatch [:navigate-back])}]
```
Comment (clauxx) [inline src/status_im/contexts/profile/settings/screens/messages/view.cljs]: can be defined outside the component

File: src/status_im/contexts/profile/settings/screens/messages/view.cljs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+   [quo/page-top {:title (i18n/label :t/messages)}]
+   [quo/category
+    {:label     (i18n/label :t/contacts)
+     :data      [{:title    (i18n/label :t/blocked-users)
+                  :on-press #(rf/dispatch [:open-modal :screen/settings-blocked-users])
```
Comment (clauxx) [inline src/status_im/contexts/profile/settings/screens/messages/view.cljs]: can be defined outside the component as well

File: src/status_im/contexts/profile/settings/screens/password/view.cljs [hunk 8]
```diff
@@ -56,31 +53,18 @@
...
-        :size                :heading-1}
-       (i18n/label :t/password)]]
+(defn view
+  []
+  (let [theme (quo.theme/use-theme-value)]
+    [quo/overlay {:type :shell :top-inset? true}
+     [quo/page-nav
+      {:background :blur
+       :icon-name  :i/arrow-left
+       :on-press   #(rf/dispatch [:navigate-back])}]
```
Comment (clauxx) [inline src/status_im/contexts/profile/settings/screens/password/view.cljs]: can be defined outside the component

File: src/status_im/subs/contact.cljs [hunk 9]
```diff
@@ -242,7 +309,16 @@
     (re-frame/subscribe [:contacts/contacts])
     (re-frame/subscribe [:profile/profile])])
  (fn [[{:keys [contacts admins]} all-contacts current-multiaccount]]
-   (contact.db/get-all-contacts-in-group-chat contacts admins all-contacts current-multiaccount)))
+   (get-all-contacts-in-group-chat contacts admins all-contacts current-multiaccount)))
+
+(defn- contact-by-address
+  [[addr contact] address]
+  (when (address/address= addr address)
```
Comment (clauxx) [inline src/status_im/subs/contact.cljs:317]: is it common to add a suffix like that in clojure?

File: src/quo/components/overlay/view.cljs [hunk 10]
```diff
@@ -2,19 +2,23 @@
...
+         :blur-radius   25
+         :blur-type     :transparent
+         :overlay-color :transparent
+         :style         style/container}
+        (into [rn/view
+               {:style (merge style/blur-container top-style container-style)}]
+              children)]
+       (into [rn/view
+              {:style (merge style/container top-style container-style)}]
+             children))]))
```
Comment (seanstrom) [inline src/quo/components/overlay/view.cljs:24]: There's an open issue for fixing the overlay component with `into`: https://github.com/status-im/status-mobile/issues/19331

Maybe we can close that issue with this PR, but it might be worth checking other usages of overlay and see if we can remove any unneeded `key` props.

File: src/status_im/subs/contact.cljs (hunk 9 above)
Comment (flexsurfer) [reply src/status_im/subs/contact.cljs:317]: i've never seen this before, please ignore this namespace it's very old code

File: src/status_im/contexts/profile/settings/screens/messages/blocked_users/view.cljs (hunk 4 above)
Comment (flexsurfer) [reply src/status_im/contexts/profile/settings/screens/messages/blocked_users/view.cljs]: leftover thanks

File: src/quo/components/list_items/user.cljs (hunk 3 above)
Comment (flexsurfer) [reply src/quo/components/list_items/user.cljs]: i'm not sure i understand, looks like not related to this PR?

File: src/status_im/contexts/profile/settings/list_items.cljs [hunk 11]
```diff
@@ -17,13 +17,12 @@
...
      :action      :arrow}]
-   [(when config/show-not-implemented-features?
-      {:title       (i18n/label :t/messages)
-       :on-press    not-implemented/alert
-       :image-props :i/messages
-       :image       :icon
-       :blur?       true
-       :action      :arrow})
+   [{:title       (i18n/label :t/messages)
+     :on-press    #(rf/dispatch [:open-modal :screen/settings-messages])
```
Comment (seanstrom) [inline src/status_im/contexts/profile/settings/list_items.cljs:21]: We can also extract this one into a static function 🙌

Comment (flexsurfer) [reply src/status_im/contexts/profile/settings/list_items.cljs:21]: done thanks

File: src/quo/components/overlay/view.cljs (hunk 10 above)
Comment (flexsurfer) [reply src/quo/components/overlay/view.cljs:24]: i think that's fine to keep keys for already added, it won't hurt I guess

File: src/status_im/contexts/contact/blocking/events.cljs [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+     {:db (assoc-in db [:contacts/contacts public-key] contact)
+      :fx [[:json-rpc/call
+            [{:method      "wakuext_blockContact"
+              :params      [public-key]
+              :js-response true
+              :on-success  (fn [^js response]
+                             (re-frame/dispatch [:contact/blocked contact (.-chats response)])
+                             (re-frame/dispatch [:sanitize-messages-and-process-response response]))
+              :on-error    #(log/error "failed to block contact" % public-key)}]]
+           [:dispatch [:hide-popover]]
```
Comment (seanstrom) [inline src/status_im/contexts/contact/blocking/events.cljs:66]: I think this was initially called inside the on-success handler when blocking a contact. Is alright to always hide the popover now?

File: src/status_im/contexts/contact/blocking/events.cljs [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+                ;; remove messages
+                (update-in [:messages chat-id] #(apply dissoc % ids))
+                (update-in [:chats chat-id]
+                           assoc
+                           :unviewed-messages-count unviewed-messages-count
+                           :unviewed-mentions-count unviewed-mentions-count
+                           :last-message            last-message))]
+    {:db (assoc-in db
+          [:message-lists chat-id]
+          (message-list/add-many nil (vals (get-in db [:messages chat-id]))))}))
```
Comment (seanstrom) [inline src/status_im/contexts/contact/blocking/events.cljs:26]: Thanks for refactoring this code from legacy 🙌 

Side note: we may want to quickly double-check that blocking/unblocking contacts from the contact profile still works correctly.

File: src/quo/components/overlay/view.cljs (hunk 10 above)
Comment (flexsurfer) [reply src/quo/components/overlay/view.cljs:24]: added issue into the description, thank you

**Review by seanstrom — COMMENTED** [review]
Looking good! 💯 
I left a few comments, and I'll come back to re-review again soon 🙏

File: src/status_im/contexts/contact/blocking/events.cljs (hunk 12 above)
Comment (flexsurfer) [reply src/status_im/contexts/contact/blocking/events.cljs:66]: yes it doesn't make sense to do not hide it, because we navigate from the screen anyway, this error may happen only in development, it can't happen in prod

File: src/status_im/contexts/contact/blocking/events.cljs (hunk 13 above)
Comment (flexsurfer) [reply src/status_im/contexts/contact/blocking/events.cljs:26]: yeah, but this code also should be refactored, but it's too complex, so it should be a separate task

**Review by clauxx — APPROVED** [review]

File: src/status_im/contexts/contact/blocking/events.cljs (hunk 13 above)
Comment (seanstrom) [reply src/status_im/contexts/contact/blocking/events.cljs:26]: Yeah absolutely! We can come back and refactor more later on 🙏

Comment (seanstrom) [reply src/status_im/contexts/contact/blocking/events.cljs:26]: I also pulled down the code and manually tested the block/unblock logic, and all is well 🙌

**Comment by qoqobolo** [comment]
@flexsurfer just found an issue here:

### ISSUE 1: The `User is unblocked` toast should be white

Screenshot 2024-04-17 at 15 00 17

Figma

Screenshot 2024-04-17 at 15 01 30

**Comment by qoqobolo** [comment]
Also, https://github.com/status-im/status-mobile/pull/19653#event-12508016154 is merged so let's rebase it and rerun tests

**Comment by flexsurfer** [comment]
@qoqobolo fixed thanks

**Comment by qoqobolo** [comment]
> @qoqobolo fixed thanks

@flexsurfer thank you! I'll relaunch e2e and will check it soon

**Comment by qoqobolo** [comment]
### ISSUE 2: Name of the previous unblocked user is replaced in the toast with the next unblocked user's name instead of showing a new toast

**Steps:**
1. Block User1 and User2
2. Go to the `Blocked users` screen 
3. Quickly unblock both users one by one paying attention on the toast

**Expected behavior:** two toasts appear one by one
**Actual behavior:** the name inside the first toast is just replaced with the second name


https://github.com/status-im/status-mobile/assets/67952253/28da7263-3b7d-4ca8-9a1a-0d9f7c6ab359

**Comment by qoqobolo** [comment]
@flexsurfer I think I found something else, could you check, please?

### ISSUE 3: Wrong `Unblock` bottom sheet color

It looks too dark both on Android and iOS. Is it possible to fix it at least for iOS (since there is a different blur-related issue on Android afaik)?

Android

Screenshot 2024-04-22 at 14 03 38

iOS

Screenshot 2024-04-22 at 14 03 59

Figma

Screenshot 2024-04-22 at 14 04 18

**Comment by flexsurfer** [comment]
@qoqobolo wondering if it's my PR or its general issue in develop with bottom sheet color in the dark mode

**Comment by qoqobolo** [comment]
> @qoqobolo wondering if it's my PR or its general issue in develop with bottom sheet color in the dark mode

I compared it with develop and the bottom sheets on the dark-themed screens that we have there, e.g. these two:

Screenshot 2024-04-23 at 11 49 32
Screenshot 2024-04-23 at 11 49 44

These are the screenshots from iOS (nightly), they look lighter.
Also, notice that the device itself is not in the dark mode, it's just the screens that only have dark theme (profile, jump to, activity center).
WDYT?

**Comment by qoqobolo** [comment]
ISSUE 3 will be discussed with the designers.
Thanks @flexsurfer, PR can be merged.
